**(a)**

We implement the **Bisection Method** as a function, bisect(). We also define the function $f$ we want to find the root for (func_1(x)). This will be used for other sections too:

In [1]:
import math

def func_1(x): 
        return(2*x + 3*math.cos(x)- math.exp(x))  # f(x) = 2x + cos(x) - e^(x)
    
def bisect(a, b, TOL, N, p_true):
    i = 1
    FA = func_1(a)
    while i <= N:
        p = a + (b-a)/2    # More accurate version for finite-digit arithmetic
        FP = func_1(p)
        print('Iteration No:', i)
        print('a =', a, ', b =', b)
        print('p =', format(p, '0.11f'))   # Show values to 11 decimal places
        print('f(p)=', FP)
        print('Absolute Error=', abs(p-p_true) ) # Absolute Error of approximation (p_true is given)
        if abs(p-p_true) <= 1e-5:
            print("NOTE: --SMALL ERROR--!")  # Throw a note once the error gets smaller than 1e-6
        if FP == 0 or (b-a)/2 < TOL:
            print("Number of iterations :", i)
            print("Approximate Solution (p) :", format(p, '0.11f'))
            break
        else:
            i += 1
        if FA*FP > 0:
            a = p
            FA = FP
        else:
            b = p
    print("CAUTION: Method Failed After", i, "Iterations!")

In [2]:
bisect(a=0, b=2, TOL=1e-6, N=100, p_true=1.23971469798)

Iteration No: 1
a = 0 , b = 2
p = 1.00000000000
f(p)= 0.9026250891453742
Absolute Error= 0.23971469798
Iteration No: 2
a = 1.0 , b = 2
p = 1.50000000000
f(p)= -1.2694774653349556
Absolute Error= 0.26028530202
Iteration No: 3
a = 1.0 , b = 1.5
p = 1.25000000000
f(p)= -0.04437587027603529
Absolute Error= 0.010285302019999998
Iteration No: 4
a = 1.0 , b = 1.25
p = 1.12500000000
f(p)= 0.46331270147796744
Absolute Error= 0.11471469798
Iteration No: 5
a = 1.125 , b = 1.25
p = 1.18750000000
f(p)= 0.2180651245349261
Absolute Error= 0.05221469798
Iteration No: 6
a = 1.1875 , b = 1.25
p = 1.21875000000
f(p)= 0.08900166110516672
Absolute Error= 0.020964697980000002
Iteration No: 7
a = 1.21875 , b = 1.25
p = 1.23437500000
f(p)= 0.022853253498943626
Absolute Error= 0.005339697980000002
Iteration No: 8
a = 1.234375 , b = 1.25
p = 1.24218750000
f(p)= -0.010626073604524056
Absolute Error= 0.002472802019999998
Iteration No: 9
a = 1.234375 , b = 1.2421875
p = 1.23828125000
f(p)= 0.00614738012727889
Abso

|Iteration Number:| $a_n$ | $b_n$ | $p_n$ | $f{(p_n)}$ | $abs(p_n - p)$  
|------------|
|1| 0 | 2 | 1.0 | 0.9026250891453742 | 0.23971469798 |
|2| 1.0 | 2 | 1.5 | -1.2694774653349556 | 0.26028530202 | 
|3| 1.0 | 1.5 | 1.25 | -0.04437587027603529 | 0.010285302019999998 |
|4| 1.0 | 1.25 | 1.125 | 0.46331270147796744 | 0.11471469798 |
|5| 1.125 | 1.25 | 1.1875 | 0.2180651245349261 | 0.05221469798 |
|6| 1.1875 | 1.25 | 1.21875 | 0.08900166110516672 | 0.020964697980000002 |
|7| 1.21875 | 1.25 | 1.234375 | 0.022853253498943626 | 0.005339697980000002 |
|8| 1.234375 | 1.25 | 1.2421875 | -0.010626073604524056 | 0.002472802019999998 |
|9| 1.234375 | 1.2421875 | 1.23828125 | 0.00614738012727889 | 0.0014334479800000022 |
|10| 1.23828125 | 1.2421875 | 1.240234375 | -0.0022308969017381663 | 0.0005196770199999978 |
|11| 1.23828125 | 1.240234375 | 1.2392578125 | 0.001960353784101976 | 0.0004568854800000022 |
|12| 1.2392578125 | 1.240234375 | 1.23974609375 | -0.00013474348008735149 | 3.1395769999997825e-05 |
|13| 1.2392578125 | 1.23974609375 | 1.23950195312 | 0.0009129371671972031 | 0.00021274485500000218 |
|14| 1.239501953125 | 1.23974609375 | 1.23962402344 | 0.00038912984791350524 | 9.067454250000218e-05 |
|15| 1.2396240234375 | 1.23974609375 | 1.23968505859 | 0.00012720143507305437 | 2.9639386250002175e-05 |
|**16**| 1.23968505859375 | 1.23974609375 | 1.23971557617 | -3.7689597083279125e-06 | **8.781918749978246e-07** |
|17| 1.23968505859375 | 1.239715576171875 | 1.23970031738 | 6.171675338073612e-05 | 1.4380597187502175e-05 |
|18| 1.2397003173828125 | 1.239715576171875 | 1.23970794678 | 2.897402576129693e-05 | 6.751202656252175e-06 |
|19| 1.2397079467773438 | 1.239715576171875 | 1.23971176147 | 1.260256525759118e-05 | 2.9365053906271754e-06 |
|20| 1.2397117614746094 | 1.239715576171875 | 1.23971366882 | 4.416810832630347e-06 | 1.0291567578146754e-06 |
|21| 1.2397136688232422 | 1.239715576171875 | 1.23971462250 | 3.239275763178284e-07 | 7.548244140842542e-08 |

**Conclusion:** After the 16th iteration the absolute error falls below $10^{-6}$. Note that the bound in part (a) can be too large as in practice it is possible to get similar accuracy with less number of iterations as is the case here.

**(b)**

We implement **Newton's Method** as a function, newton_raph() in a similar way. In addition, we derive $f'$ of the function $f$ and define it as f_prime(x) :

In [3]:
def f_prime(x):       # Derivative of the function defined 
    return 2 - 3*math.sin(x) - math.exp(x)    # f'(x) = 2 - 3sin(x) - e^(x)

def newton_raph(p0, TOL, N, p_true):
    i = 1
    while i <= N:
        p = p0 - (func_1(p0)/f_prime(p0))
        print('Iteration No:', i)
        print('p=', format(p, '0.11f'))
        print('f(p)=', func_1(p))
        print('f\'(p)=', f_prime(p))
        print('Absolute Error=', abs(p-p_true))
        if abs(p-p_true) < 1e-6:
            print("NOTE: --SMALL ERROR--!") # Throw a note once the error gets smaller than 1e-6
        if abs(p-p0) < TOL:
            print("Number of iterations :", i)
            print("Approximate Solution (p) :", format(p, '0.11f'))
            break
        else:
            i += 1
        p0 = p
    print("CAUTION: Method Failed After", i, "Iterations!")

In [4]:
newton_raph(p0=2, TOL=1e-20, N=100, p_true=1.23971469798)

Iteration No: 1
p= 1.42866501155
f(p)= -0.8908345999805163
f'(p)= -5.142873409873063
Absolute Error= 0.18895031357321113
Iteration No: 2
p= 1.25544772056
f(p)= -0.06807009839109623
f'(p)= -4.361474208902822
Absolute Error= 0.015733022579749845
Iteration No: 3
p= 1.23984058978
f(p)= -0.0005403251438433898
f'(p)= -4.292259283300442
Absolute Error= 0.00012589179535837935
Iteration No: 4
p= 1.23971470615
f(p)= -3.509944157542577e-08
f'(p)= -4.291701635862026
Absolute Error= 8.17366152539023e-09
NOTE: --SMALL ERROR--!
Iteration No: 5
p= 1.23971469798
f(p)= -8.881784197001252e-16
f'(p)= -4.291701599632942
Absolute Error= 4.782396700875324e-12
NOTE: --SMALL ERROR--!
Iteration No: 6
p= 1.23971469798
f(p)= 0.0
f'(p)= -4.291701599632942
Absolute Error= 4.782618745480249e-12
NOTE: --SMALL ERROR--!
Iteration No: 7
p= 1.23971469798
f(p)= 0.0
f'(p)= -4.291701599632942
Absolute Error= 4.782618745480249e-12
NOTE: --SMALL ERROR--!
Number of iterations : 7
Approximate Solution (p) : 1.23971469798
CAUTIO

|Iteration Number:| $p_n$ | $f{(p_n)}$ | $f'{(p_n)}$ | $abs(p_n - p)$ | 
|------------|
|0| 2 | -4.637496608572078 | -8.116948379407695 | 0.76028530202 |
|1| 1.42866501155 | -0.8908345999805163 | -5.142873409873063 | 0.18895031357321113 | 
|2| 1.25544772056 | -0.06807009839109623 | -4.361474208902822 | 0.015733022579749845 | 
|3| 1.23984058978 | -0.0005403251438433898 | -4.292259283300442 | 0.00012589179535837935 | 
|**4**| 1.23971470615 | -3.509944157542577e-08 | -4.291701635862026 | **8.17366152539023e-09** | 
|5| 1.23971469798 | -8.881784197001252e-16 | -4.291701599632942 | 4.782396700875324e-12 | 
|6| 1.23971469798 | 0.0 | -4.291701599632942 | 4.782618745480249e-12 | 
|7| 1.23971469798 | 0.0 | -4.291701599632942 | 4.782618745480249e-12 | 

**Conclusion:** After the 4th iteration the absolute error falls below  $10^{−6}$. This means we have faster convergence compared to the Bisection Method of part (a). The identical results for $p_n$ in iteraions 6 and 7 show that the method has converged to the 'approximately true' value of the root.

**(c)**

We implement the **Secant Method** as a function, secant(), in a similar way :

In [5]:
def secant(p0, p1, TOL, N, p_true):
    i = 2
    q0 = func_1(p0)
    q1 = func_1(p1)
    print('Iteration No : 0 \n p=', format(p0, '0.11f'))
    print('Iteration No : 1 \n p=', format(p1, '0.11f'))
    while i <= N :
        p = p1 - (q1*((p1-p0)/(q1-q0)))
        print('Iteration No:', i)
        print('p=', format(p, '0.11f'))
        print('f(p)=', func_1(p))
        print('Absolute Error=', abs(p-p_true))
        if abs(p-p_true) < 1e-6:
            print("NOTE: --SMALL ERROR--!") # Throw a note once the error gets smaller than 1e-6
        if abs(p-p0) < TOL:
            print("Number of iterations :", i)
            print("Approximate Solution (p) :", format(p, '0.11f'))
            break
        else:
            i += 1
        p0 = p1
        q0 = q1
        p1 = p
        q1 = func_1(p)
    print("CAUTION: Method Failed After", N, "Iterations!")

In [6]:
secant(p0=10 ,p1=2 ,TOL=1e-7, N=100, p_true=1.23971469798)

Iteration No : 0 
 p= 10.00000000000
Iteration No : 1 
 p= 2.00000000000
Iteration No: 2
p= 1.99831397063
f(p)= -4.623819919413408
Absolute Error= 0.7585992726469828
Iteration No: 3
p= 1.42830056247
f(p)= -0.8889605898239727
Absolute Error= 0.18858586448737147
Iteration No: 4
p= 1.29262759307
f(p)= -0.23330369295699382
Absolute Error= 0.05291289509046426
Iteration No: 5
p= 1.24435081150
f(p)= -0.01994443247808153
Absolute Error= 0.004636113520725793
Iteration No: 6
p= 1.23983798647
f(p)= -0.0005291511163671991
Absolute Error= 0.0001232884944508683
Iteration No: 7
p= 1.23971499231
f(p)= -1.2631769212845256e-06
Absolute Error= 2.943252634945992e-07
NOTE: --SMALL ERROR--!
Iteration No: 8
p= 1.23971469799
f(p)= -8.036904475261508e-11
Absolute Error= 1.3943957100082116e-11
NOTE: --SMALL ERROR--!
Iteration No: 9
p= 1.23971469798
f(p)= 0.0
Absolute Error= 4.782618745480249e-12
NOTE: --SMALL ERROR--!
Iteration No: 10
p= 1.23971469798
f(p)= 0.0
Absolute Error= 4.782618745480249e-12
NOTE: --SMAL

|Iteration Number:| $p_n$ | $f{(p_n)}$ | $abs(p_n - p)$ |
|------------|
|0| 10 | -22008.98300939395 | 8.76028530202 | 
|1| 2 | -4.637496608572078 | 0.76028530202 | 
|2| 1.99831397063 | -4.623819919413408 | 0.7585992726469828 | 
|3| 1.42830056247 | -0.8889605898239727 | 0.18858586448737147 | 
|4| 1.29262759307 | -0.23330369295699382 | 0.05291289509046426 |
|5| 1.24435081150 | -0.01994443247808153 | 0.004636113520725793 |
|6| 1.23983798647 | -0.0005291511163671991 | 0.0001232884944508683 |
|**7**| 1.23971499231 | -1.2631769212845256e-06 | **2.943252634945992e-07** |
|8| 1.23971469799 | -8.036904475261508e-11 | 1.3943957100082116e-11 |
|9| 1.23971469798 | 0.0 | 4.782618745480249e-12 |
|10| 1.23971469798 | 0.0 | 4.782618745480249e-12 |

**Conclusion:** After the 7th iteration the absolute error falls below  $10^{−6}$. This means we have faster convergence compared to the Bisection Method of part (a) but slightly slower convergence compared to the Newton Method of part (b). The identical results for $p_n$ in iteraions 9 and 10 show that the method has converged to the 'approximately true' value of the root by the 9th iteration (we could have stopped there).

---------

---------

**(a)**

First we define $f(h)$ as a function func_2(x). Next we implement the **Method of False Position (Regula Falsi)** as a function false_position() :

In [7]:
def func_2(x):
    return 5*(math.pi) - 10*(math.asin(x)) - (10*x)*(math.sqrt(1-x**2)) - 12    # f(x) = 5(pi) - 10arcsin(x) - 10x.sqrt(x^2) -12(= 0)

def false_position(p0, p1, TOL, N):
    i = 2
    q0 = func_2(p0)
    q1 = func_2(p1)
    while i <= N :
        p = p1 - (q1*((p1-p0)/(q1-q0)))
        if i == 2 :
            print('f(p0)=', format(func_2(p0), '0.11f'))
            print('f(p1)=', format(func_2(p1), '0.11f'))
        print('Iteration No:', i-1)
        print('p0=', p0)
        print('p1=', p1)
        print('p=', format(p, '0.11f'))
        print('f(p)=', func_2(p))
        if abs(func_2(p)) < TOL :    # The stopping criterion from part (a) - given: 0.1
            print("Number of iterations :", i-1)
            print("Approximate Solution (p) :", format(p, '0.11f'))
            break
        else:
            i += 1
            q = func_2(p)
        if q*q1 < 0:
            p0 = p1
            q0 = q1
        p1 = p
        q1 = q
    print("CAUTION: Method Failed After", N, "Iterations!")

In [8]:
false_position(p0= 0 ,p1= 1,TOL= 1e-1, N=100)

f(p0)= 3.70796326795
f(p1)= -12.00000000000
Iteration No: 1
p0= 0
p1= 1
p= 0.23605627316
f(p)= -0.9689426839999342
Iteration No: 2
p0= 0
p1= 0.23605627315890243
p= 0.18715107788
f(p)= -0.013091814901947885
Number of iterations : 2
Approximate Solution (p) : 0.18715107788
CAUTION: Method Failed After 100 Iterations!


|Iteration Number:| $p_0$ | $p_1$ | $p$ | $f(p)$ | 
|------------|
|1| 0 | 1 | 0.23605627316 | -0.9689426839999342 |
|2| 0 | 0.23605627315890243 | **0.18715107788** | -0.013091814901947885 |